### PR TITLE
Prometheus: Fix exemplars not respecting corresponding series display status.

### LINKED ISSUE #59678: Prometheus: Exemplar queries with multiple series in a single metric are not differentiable in the UI
**What happened**:

Exemplars in the prometheus graphs do not toggle when a user has selected a specific label to view.
That is, f you click on a series in a graph, and all of the other series are no longer rendered, you'd expect that the exemplars from other series would be hidden as well. This makes working with exemplars on panels with many series extremely difficult. 

image
You can see in the screenshot above that only the `tns/loadgen` app is being displayed, and yet an exemplar from `tns/app` is being displayed. 

**What you expected to happen**:
Exemplars would toggle with their associated series.

**How to reproduce it (as minimally and precisely as possible)**:
Create a timeseries graph with exemplars and group it by an additional label: e.g. `histogram_quantile(0.95, sum(rate(tns_request_duration_seconds_bucket[$__rate_interval])) by (le, job))`

**Anything else we need to know?**:
Exemplars are complicated, I'm still looking into this and will try to keep this issue updated as I learn more.

**Environment**:
- Grafana version: 9.x
- Data source type & version: Prometheus & all
- OS Grafana is installed on: All
- User OS & Browser: All
- Grafana plugins: N/A

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Fix bug in which exemplars do not filter when a series is hidden via user clicking on the legend UI.

Currently, when filtering a metric with multiple labels/series via the legend UI, the exemplars do not respect the active filters, and will always display all exemplars in the current time-series graph.

image

image

image

**Why do we need this feature?**
Users with many labels for a single series in an exemplar query are left with a heap of exemplars that are not differentiated. This PR allows users to hide series from the graph and only view the relevant exemplars.

**Who is this feature for?**
Users of prometheus exemplars.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/59678

**Special notes for your reviewer**:
This might not be the most clean or efficient solution, please let me know if there's any improvements we can make to this!

Debug snapshot from the TNS app:
`histogram_quantile(0.95, sum by(le, job) (rate(tns_request_duration_seconds_bucket[$__rate_interval])))`
[debug-Exemplar 1 metric 3 labels-2022-12-07 11_53_14.json.txt](https://github.com/grafana/grafana/files/10178658/debug-Exemplar.1.metric.3.labels-2022-12-07.11_53_14.json.txt)

### CODE REVIEW

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx [hunk 1]
```diff
@@ -19,9 +20,40 @@ interface ExemplarsPluginProps {
...
+  const visibleLabels: Labels[] = [];
+  if (frames?.length) {
+    visibleSeries.forEach((plotInstance) => {
+      const frameIndex = plotInstance.props?.dataFrameFieldIndex?.frameIndex;
+      const fieldIndex = plotInstance.props?.dataFrameFieldIndex?.fieldIndex;
+
+      if (frameIndex !== undefined && fieldIndex !== undefined) {
+        const field = frames[frameIndex].fields[fieldIndex];
+        if (field.labels) {
+          // Note that this may be an empty object
```
Comment (gtk-grafana) [inline public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:40]: whyyyyy

**Comment by gtk-grafana** [comment]
The e2e test cases we currently have are only testing a single series, we'll need to revisit our exemplar e2e tests in another issue: https://github.com/grafana/grafana/issues/59911.

**Comment by gtk-grafana** [comment]
@gtk-grafana We might want to add change log notes?

File: public/app/plugins/panel/candlestick/CandlestickPanel.tsx [hunk 2]
```diff
@@ -313,6 +313,7 @@ export const CandlestickPanel: React.FC<CandlestickPanelProps> = ({
             )}
             {data.annotations && (
               <ExemplarsPlugin
+                visibleLabels={getVisibleLabels(config, [info.frame])}
```
Comment (gtk-grafana) [inline public/app/plugins/panel/candlestick/CandlestickPanel.tsx:316]: Is this necessary or should this property be optional? What does it do in this context?

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx [hunk 3]
```diff
@@ -61,8 +69,41 @@ export const ExemplarsPlugin: React.FC<ExemplarsPluginProps> = ({ exemplars, tim
...
+    Object.values(labels).forEach((labelObject) => {
+      return Object.keys(labelObject).forEach((labelName) => {
+        const labelValue: string = labelObject[labelName];
+        if (typeof labelToSet[labelName] === 'undefined') {
+          labelToSet[labelName] = new Set<string>().add(labelValue);
+        } else {
+          labelToSet[labelName].add(labelValue);
+        }
+      });
+    });
```
Comment (itsmylife) [inline public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:85]:
```suggestion
    const labelToSet: Record<string, Set<string>> = {};

    Object.values(labels).forEach((labelObject) => {
      return Object.keys(labelObject).forEach((labelName) => {
        const labelValue: string = labelObject[labelName];
        if (!labelToSet[labelName]) {
          labelToSet[labelName] = new Set<string>();
        }
        labelToSet[labelName].add(labelValue);
      });
    });
```

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx [hunk 4]
```diff
@@ -61,8 +69,41 @@ export const ExemplarsPlugin: React.FC<ExemplarsPluginProps> = ({ exemplars, tim
...
+      showMarker = showExemplarMarker(
+        visibleLabels,
+        showMarker,
+        getUniqueValuesFromLabels,
+        dataFrame,
+        dataFrameFieldIndex
+      );
+
+      if (!showMarker) {
+        return <></>;
```
Comment (itsmylife) [inline public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:104]: we can also return null but this is also fine.

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx (hunk 3 above)
Comment (itsmylife) [reply public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:85]: Also, I do not understand why we need to return inside `forEach` in line 77. I think there is no need to return. Am I missing something?

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx [hunk 5]
```diff
@@ -86,3 +127,69 @@ export const ExemplarsPlugin: React.FC<ExemplarsPluginProps> = ({ exemplars, tim
...
+export const getVisibleLabels = (
+  config: UPlotConfigBuilder,
+  frames: DataFrame[] | null
+): { labels: Labels[]; totalSeriesCount: number } => {
+  const visibleSeries = config.series.filter((series) => series.props.show);
+  const visibleLabels: Labels[] = [];
+  if (frames?.length) {
+    visibleSeries.forEach((plotInstance) => {
+      const frameIndex = plotInstance.props?.dataFrameFieldIndex?.frameIndex;
+      const fieldIndex = plotInstance.props?.dataFrameFieldIndex?.fieldIndex;
```
Comment (itsmylife) [inline public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:143]: Would they return undefined or `NaN` if optional chaining fails without reaching to `fieldIndex` or `frameIndex`

**Review by itsmylife — COMMENTED** [review]
In general, it looks good to me. I left some nitpicks.
It would be nice if we have some unit test coverage for the methods you introduced in `ExemplarsPlugin`.

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx (hunk 5 above)
Comment (gtk-grafana) [reply public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx:138]: should be [undefined](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)

File: public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx (hunk 3 above)
Comment (gtk-grafana) [reply public/app/plugins/panel/timeseries/plugins/ExemplarsPlugin.tsx]: This function isn't even being called, I'm removing it now.

**Comment by gtk-grafana** [comment]
@itsmylife thanks for the feedback, I've implemented a unit test for the exported function. Can you take another look in the morning?

**Review by bohandley — APPROVED** [review]
Makes sense to me, well done! LGTM!

**Review by itsmylife — APPROVED** [review]
LGTM but please revert the changes in `CHANGELOG.md` file.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,12 @@
+<!-- 9.3.2 START -->
+
+# 9.3.2 (2022-12-13)
+
+### Bug fixes
+
+- **Prometheus:** Fix exemplars not respecting corresponding series display status. [#59678](https://github.com/grafana/grafana/issues/59678), [@gtk-grafana](https://github.com/gtk-grafana)
+
+<!-- 9.3.2 END -->
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply CHANGELOG.md:9]: Yup sorry I forgot how this works 🤦

**Pull request merged**